test/drain_test.go

```diff
@@ -0,0 +1,296 @@
+
+	num := 100
+	subs := make([]*nats.Subscription, num)
+
+	// Normal Unsubscribe
+	for i := 0; i < num; i++ {
+		sub, err := nc.Subscribe("foo", func(_ *nats.Msg) {})
+		if err != nil {
+			t.Fatalf("Error creating subscription; %v\n", err)
+		}
+		subs[i] = sub
+	}
+
+	if numSubs := nc.NumSubscriptions(); numSubs != num {
+		t.Fatalf("Expected %d subscriptions, got %d\n", num, numSubs)
+	}
+	for i := 0; i < num; i++ {
+		subs[i].Unsubscribe()
+	}
+	if numSubs := nc.NumSubscriptions(); numSubs != 0 {
+		t.Fatalf("Expected no subscriptions, got %d\n", numSubs)
+	}
+
+	// Drain version
+	for i := 0; i < num; i++ {
+		sub, err := nc.Subscribe("foo", func(_ *nats.Msg) {})
+		if err != nil {
+			t.Fatalf("Error creating subscription; %v\n", err)
+		}
+		subs[i] = sub
+	}
+
+	if numSubs := nc.NumSubscriptions(); numSubs != num {
+		t.Fatalf("Expected %d subscriptions, got %d\n", num, numSubs)
+	}
+	for i := 0; i < num; i++ {
+		subs[i].Drain()
+	}
+	// Should happen quickly that we get to zero, so do not need to wait long.
+	waitFor(t, 2*time.Second, 10*time.Millisecond, func() error {
+		if numSubs := nc.NumSubscriptions(); numSubs != 0 {
+			return fmt.Errorf("Expected no subscriptions, got %d\n", numSubs)
+		}
+		return nil
+	})
+}
+
+func TestDrainSlowSubscriber(t *testing.T) {
+	s := RunDefaultServer()
+	defer s.Shutdown()
+	nc := NewDefaultConnection(t)
+	defer nc.Close()
+
+	received := int32(0)
+
+	sub, err := nc.Subscribe("foo", func(_ *nats.Msg) {
+		atomic.AddInt32(&received, 1)
+		time.Sleep(100 * time.Millisecond)
+	})
+	if err != nil {
+		t.Fatalf("Error creating subscription; %v\n", err)
+	}
+
+	total := 10
+	for i := 0; i < total; i++ {
+		nc.Publish("foo", []byte("Slow Slow"))
+	}
+	nc.Flush()
+
+	pmsgs, _, _ := sub.Pending()
+	if pmsgs != total && pmsgs != total-1 {
+		t.Fatalf("Expected most messages to be pending, but got %d vs %d\n", pmsgs, total)
+	}
+	sub.Drain()
+
+	// Should take a second or so to drain away.
+	waitFor(t, 2*time.Second, 100*time.Millisecond, func() error {
+		// Wait for it to become invalid. Once drained it is unsubscribed.
+		_, _, err := sub.Pending()
+		if err != nats.ErrBadSubscription {
+			return fmt.Errorf("Still valid")
+		}
+		r := int(atomic.LoadInt32(&received))
+		if r != total {
+			t.Fatalf("Did not receive all messages, got %d vs %d\n", r, total)
+		}
+		return nil
+	})
+}
+
+func TestDrainConnection(t *testing.T) {
+	s := RunDefaultServer()
+	defer s.Shutdown()
+
+	done := make(chan bool)
+
+	closed := func(nc *nats.Conn) {
+		done <- true
+	}
+
+	url := fmt.Sprintf("nats://127.0.0.1:%d", nats.DefaultPort)
+	nc, err := nats.Connect(url, nats.ClosedHandler(closed))
+	if err != nil {
+		t.Fatalf("Failed to create default connection: %v\n", err)
+	}
+	defer nc.Close()
+
+	received := int32(0)
+	expected := int32(50)
+	sleep := 10 * time.Millisecond
+
+	// Create a slow subscriber
+	sub, err := nc.Subscribe("foo", func(_ *nats.Msg) {
+		time.Sleep(sleep)
+		atomic.AddInt32(&received, 1)
```

Comment:
Could we add to this test a subscription on "bar" that this callback sends a message to? This way, we would check that publishes are ok and ensure that all messages on bar are received too.

Reply:
fixed